kmq.js

```diff
@@ -11,34 +11,86 @@ const db = new sqlite3.Database('./main.db', (err) => {
         return;
     }
 });
-var scoreboard = {};
+var scoreboard = [];
 var currentSong = null;
 var currentArtist = null;
+var currentSongLink = null;
 var gameInSession = false;
+
+const songInfoMessage = (message, currentSong, currentArtist, currentSongLink, isQuit) => {
+    message.channel.send({embed: {
+        color: 15158332, // Red
+        author: {
+            name: isQuit ? null : message.author.username,
+            icon_url: isQuit ? null : message.author.avatarURL
+        },
+        title: currentSong + " - " + currentArtist,
+        description: "https://youtube.com/watch?v=" + currentSongLink,
+        image: {
+            url: "https://img.youtube.com/vi/" + currentSongLink + "/hqdefault.jpg"
+        }
+    }})
+}
+
+const sendScoreboard = (message, scoreboard) => {
+    message.channel.send({embed: {
+        color: 15158332,
+        title: "**Results**",
+        fields: scoreboard
+    }})
+}
+
 client.on('ready', () => {
     console.log(`Logged in as ${client.user.tag}!`);
 });
+
 client.on('message', message => {
     if (message.author.equals(client.user)) return;
     let command = parseCommand(message.content) || null;
     if (command) {
         if (command.action === "stop") {
             gameInSession = false;
-            message.channel.send("The correct song was: " + currentSong + " by " + currentArtist);
+            songInfoMessage(message, currentSong, currentArtist, currentSongLink, true);
             disconnectVoiceConnection(message);
         }
         else if (command.action === "random") {
             startGame(message);
         }
+        else if (command.action === "end") {
+            if (scoreboard.length > 0) {
+                disconnectVoiceConnection(message);
+                message.channel.send(scoreboard[0].name + " wins!");
+                sendScoreboard(message, scoreboard);
+                scoreboard = [];
+            }
+        }
     }
     else {
         let guess = cleanSongName(message.content);
         if (currentSong && guess === cleanSongName(currentSong)) {
             //this should be atomic
-            scoreboard[getUserIdentifier(message.author)] = (scoreboard[getUserIdentifier(message.author)] || 0) + 1;
-            message.channel.send("Correct answer was: " + currentSong + " by " + currentArtist + "\n" + JSON.stringify(scoreboard));
+            let index = -1;
+            for (let i = 0; i < scoreboard.length; i++) {
```

Comment:
Instead of having scoreboard as an array and looking up each user in O(n), you can just have an object with each user identifier as a key in O(1)